# Load and normlize methylation data
## ACC - JÃ¤kel et al.
## by AVV
### Bronnen: https://www.bioconductor.org/packages/release/workflows/vignettes/methylationArrayAnalysis/inst/doc/methylationArrayAnalysis.html#filtering en https://htmlpreview.github.io/?https://github.com/hansenlab/tutorial.450k/blob/master/inst/doc/methylation450k.html#batch-effects-correction-with-sva

#####################################
########### Overview ################
#####################################
1. Load required packages
2. Read data
3. Create classes: MethylSet and RadioSet
4. Quality control
5. Normalization
6. Filtering of probes
7. Save data

#######################################
####### 1. Load required packages #####
#######################################
```{r}
library(minfi)
library(minfiData)
library(conumee)
library(limma)
library(RColorBrewer)
library(shinyMethyl)
```
## Check directory
```{r}
getwd()
```
#####################################
########## 2. Read data #############
#####################################
# Parsing IDAT files from Illumina methylation arrays.
output = RGSET: contains raw intensities in the green and red channels, accompanied with pheno data
```{r}
baseDir <- "T:/pathologie/PRL/Groep-Brosens/2. Anna Vera/2. ACN-SPN-NET/Pancreas-ID/00_Methylation Data/2017_Jakel_450K_ACC, PanNET, MacNEC, Normal Pancreas, Pancreatic Cells/IdatFiles_ACC"
targets <- read.metharray.sheet(baseDir)
RGSet <- read.metharray.exp(base = baseDir, targets = targets, verbose=T)
```
## To aces pheno data use:
```{r}
phenoData <- pData(RGSet)
phenoData[,1:6]
```
## The RGChannelSet stores also a manifest object that contains the probe design information of the array:
```{r}
manifest <- getManifest(RGSet)
manifest
```
##########################################
####### create MethylSet and RadioSet#####
##########################################

##Create Methylset & RdioSet
```{r}
MSet <- preprocessRaw(RGSet) 
RSet <- ratioConvert(MSet, what = "both", keepCN = TRUE)
```
## return the Beta value matrix, M value matrix and the Copy Number matrix.
```{r}
beta <- getBeta(RSet)
M <- getM(RSet)
CN <- getCN(RSet)
```

#Create Genomic RadioSet
```{r}
GRset <- mapToGenome(RSet)
```
## functions to acces data
```{r}
beta <- getBeta(GRset)
M <- getM(GRset)
CN <- getCN(GRset)
sampleNames <- sampleNames(GRset)
probeNames <- featureNames(GRset)
pheno <- pData(GRset)
```
## return probe locations:
````{r}
gr <- granges(GRset)
head(gr, n=3)
````
## Access the full annotation:
```{r}
annotation <- getAnnotation(GRset)
names(annotation)
```

##################################
####### 4. Quality control #######
##################################

When plotting the medians of methylated and umethylated data intensities good samples cluster together, bad samples tend to separate and have owe median intensities.

The methylated and unmethylated signals are already converted into MethylSet (MSet)

## calculate mean P-values
We can generate a detection p-value for every CpG in every sample, which is indicative of the quality of the signal. The method used by minfi to calculate detection p-values compares the total signal (M+U) for each probe to the background signal level, which is estimated from the negative control probes. Very small p-values are indicative of a reliable signal whilst large p-values, for example >0.01, generally indicate a poor quality signal.
Plotting the mean detection p-value for each sample allows us to gauge the general quality of the samples in terms of the overall signal reliability. Samples that have many failed probes will have relatively large mean detection p-values.
```{r}
Pvalue <- detectionP(RGSet)

pal <- brewer.pal(8,"Dark2")
par(mfrow=c(1,2))
barplot(colMeans(Pvalue), col=pal[factor(targets$Sample_Group)], las=2, 
        cex.names=0.8, ylab="Mean detection p-values")
abline(h=0.05,col="red")
legend("topleft", legend=levels(factor(targets$Sample_Group)), fill=pal,
       bg="white")

barplot(colMeans(Pvalue), col=pal[factor(targets$Sample_Group)], las=2, 
        cex.names=0.8, ylim=c(0,0.002), ylab="Mean detection p-values")
abline(h=0.05,col="red")
legend("topleft", legend=levels(factor(targets$Sample_Group)), fill=pal, 
       bg="white")
```
## to save plots:
```{r}
pdf("T:/pathologie/PRL/Groep-Brosens/2. Anna Vera/2. ACN-SPN-NET/Pancreas-ID/01_Load&Normalize/Quality Control/P-values_ACC_Jakel.pdf")
par(mfrow=c(1,2))
barplot(colMeans(Pvalue), col=pal[factor(targets$Sample_Group)], las=2, 
        cex.names=0.8, ylab="Mean detection p-values")
abline(h=0.05,col="red")
legend("topleft", legend=levels(factor(targets$Sample_Group)), fill=pal,
       bg="white")
barplot(colMeans(Pvalue), col=pal[factor(targets$Sample_Group)], las=2, 
        cex.names=0.8, ylim=c(0,0.002), ylab="Mean detection p-values")
abline(h=0.05,col="red")
legend("topleft", legend=levels(factor(targets$Sample_Group)), fill=pal, 
       bg="white")
dev.off()
``` 

## Remove poor quality samples:
```{r}
# remove poor quality samples
keep1 <- colMeans(Pvalue) < 0.05
RGSet1 <- RGSet[,keep1]
# remove poor quality samples from targets data
targets1 <- targets[keep1,]
targets1[,1:5]
table(keep1)
# remove poor quality samples from detection p-value table
detP <- Pvalue[,keep1]
dim(detP)
```

```{r}
qc <- getQC(MSet)
MSetEx <- addQC(MSet, qc = qc)
plotQC(qc)
```
## Look at Beta value densities with densityplot or density beanplot
```{r}
densityPlot(MSet, sampGroups = phenoData$Sample_Group)
densityBeanPlot(MSet, sampGroups = phenoData$Sample_Group)
````
## Probe control
```{r}
controlStripPlot(RGSet, controls="BISULFITE CONVERSION II")
```
## create QC report:
```{r}
qcReport(RGSet, 
         sampNames = phenoData$listData$Sample_Name, 
         sampGroups = phenoData$listData$Sample_Group, 
         pdf= "T:/pathologie/PRL/Groep-Brosens/2. Anna Vera/2. ACN-SPN-NET/Pancreas-ID/01_Load&Normalize/Quality Control/QC-report_ACC_Jakel.pdf")
```
### Sex prediction
```{r}
predictedSex <- getSex(GRset, cutoff = -2)$predictedSex
table(predictedSex)
```
###############################
####### 5.normalization #######
###############################

```{r}
set.seed(666)
MSet.ssnoob <- preprocessNoob(RGSet, dyeMethod= "single")
betas.ssnoob.ACCJakel <- getBeta(MSet.ssnoob)
````
## to save data
```{r}
save(MSet.ssnoob, file = 'T:/pathologie/PRL/Groep-Brosens/2. Anna Vera/2. ACN-SPN-NET/Pancreas-ID/01_Load&Normalize/PreProcessedDataSets/nonfilteredMSet.noobPCJakel17042022.rData')
save(betas.ssnoob.ACCJakel, file = 'T:/pathologie/PRL/Groep-Brosens/2. Anna Vera/2. ACN-SPN-NET/Pancreas-ID/01_Load&Normalize/PreProcessedDataSets/nonfilteredbetas.noobPCJakel17042022.rData')

save(betas.ssnoob.ACCJakel, file = 'T:/pathologie/PRL/Groep-Brosens/2. Anna Vera/2. ACN-SPN-NET/Pancreas-ID/01_Load&Normalize/PreProcessedDataSets/nonfilteredbetas.noobPCJakel17042022.rds')
```
# effect of normalisation:
```{r}
mdsPlot(getM(MSet.ssnoob), 
        numPositions = 1000, 
        sampGroups = targets$status,
        sampNames = targets$person,
        legendPos = "bottomleft",
        main = "preprocess ssNoob")
dev.off()
```
## visualisation before after normalisation
```{R}
# visualise what the data looks like before and after normalization
par(mfrow=c(1,2))
pdf("T:/pathologie/PRL/Groep-Brosens/2. Anna Vera/2. ACN-SPN-NET/Pancreas-ID/01_Load&Normalize/Quality Control/Normalization_ACC_Jakel.pdf")
densityPlot(RGSet, sampGroups=targets$Sample_Group,main="Raw", legend=FALSE)
legend("top", legend = levels(factor(targets$Sample_Group)), 
       text.col=brewer.pal(8,"Dark2"))
densityPlot(getBeta(MSet.ssnoob), sampGroups=targets$Sample_Group,
            main="Normalized", legend=FALSE)
legend("top", legend = levels(factor(targets$Sample_Group)), 
       text.col=brewer.pal(8,"Dark2"))
dev.off()
```
## ShinyData
```{r}
summary <- shinySummarize(RGSet)
RSet.ssnoob <- ratioConvert(MSet.ssnoob, what = "both", keepCN = TRUE)
GRSet.ssnoob <- mapToGenome(RSet.ssnoob)
summary.Noob <- shinySummarize(GRSet.ssnoob)
```

```{r}
runShinyMethyl(summary, summary.Noob)
```

############################################
############ Filter probes #################
############################################


```{r}
detP <- detectionP(RGSet)
```

```{r}
# transform MSet to GRSet
RSet.ssnoob <- ratioConvert(MSet.ssnoob, what = "both", keepCN = TRUE)
GRSet.ssnoob <- mapToGenome(RSet.ssnoob)
annotation <- getAnnotation(GRSet.ssnoob)
# ensure probes are in the same order in the GRset.quantile and Pvalue objects
detP1 <- detP[match(featureNames(GRSet.ssnoob),rownames(detP)),] 
dim(detP1)

# remove any probes that have failed in one or more samples
keep2 <- rowSums(detP1 < 0.01) == ncol(GRSet.ssnoob) 
table(keep2)

GRSet.ssnoobFlt <- GRSet.ssnoob[keep2,]
GRSet.ssnoobFlt
```
## Filter XY chromosomes
```{r}
# if your data includes males and females, remove probes on the sex chromosomes
keep3 <- !(featureNames(GRSet.ssnoobFlt) %in% annotation$Name[annotation$chr %in% 
                                                        c("chrX","chrY")])
table(keep3)
GRSet.ssnoobFltXY <- GRSet.ssnoobFlt[keep3,]
GRSet.ssnoobFltXY
```
## remove probes with SNPs at CpG site
```{r}
GRSet.ssnoobFltXYSNP <- dropLociWithSnps(GRSet.ssnoobFltXY)
GRSet.ssnoobFltXYSNP
```
## remove cross-active probes
```{r}
xReactiveProbes <- read.csv("https://github.com/sirselim/illumina450k_filtering/blob/master/48639-non-specific-probes-Illumina450k.csv", stringsAsFactors=FALSE)
keep4 <- !(featureNames(GRSet.ssnoobFltXYSNP) %in% xReactiveProbes$TargetID)
table(keep4)
GRSet.ssnoobFltXYSNPX <- GRSet.ssnoobFltXYSNP[keep4,] 
GRSet.ssnoobFltXYSNPX
```

## calculate B-values for statistical analysis
```{r}
bValsnoob <- getBeta(GRSet.ssnoobFltXYSNPX)
head(bValsnoob[,1:5])
```

## Save normalized data for later use
```{r}
bValsnoobdf <- as.data.frame(bValsnoob)
head(bValsnoobdf[1:10,1:10])
save(bValsnoobdf, file = 'T:/pathologie/PRL/Groep-Brosens/2. Anna Vera/2. ACN-SPN-NET/Pancreas-ID/01_Load&Normalize/PreProcessedDataSets/preprocessedssNoobACCJakel11032022.rData')
```